### PR TITLE
Front end for editing species

### PR DESCRIPTION
**Pull request by anissa-agahchen**
# Overview

## This PR contains the following changes

Front -end for editing the Species Component

### General

- [ ] I have performed a self-review of my own code

### Code

- [ ] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [ ] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [ ] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [ ] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [ ] My code changes account for null/undefined values and handle errors appropriately
- [ ] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [ ] My code follows the established style conventions
- [ ] My code uses native material-ui components/icons/conventions when possible

### Documentation

- [ ] I have commented my code sufficiently, such that an unfamiliar developer could understand my code
- [ ] I have added/updated README's and related documentation, as needed

### Tests

- [ ] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [ ] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [ ] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Screenshots

Please add any relevant UI screenshots, if applicable.

### CODE REVIEW

File: api/src/queries/project/project-view-queries.test.ts [hunk 1]
```diff
@@ -13,6 +11,8 @@ import {
   getRegionsByProjectSQL
 } from './project-view-queries';
 
+import { getFocalSpeciesByProjectSQL, getAncillarySpeciesByProjectSQL } from './project-view-update-queries';
```
Comment (NickPhura) [inline api/src/queries/project/project-view-queries.test.ts]: These are now being tested in the project-view-update-queries.test.ts, so we can remove duplicate tests from this file.

File: app/src/features/projects/components/ProjectSpeciesForm.tsx [hunk 2]
```diff
@@ -4,7 +4,7 @@ import MultiAutocompleteFieldVariableSize, {
 } from 'components/fields/MultiAutocompleteFieldVariableSize';
 import { useFormikContext } from 'formik';
 import React from 'react';
-import yup from 'utils/YupSchema';
+import * as yup from 'yup';
```
Comment (NickPhura) [inline app/src/features/projects/components/ProjectSpeciesForm.tsx]: Can we revert back to the previous version (where yup is imported from the `utils/YupSchema`.  It shouldn't require any changes to anything below, it just has some of the extended yup tests I added.

**Comment by NickPhura** [comment]
Looks good, just the 2 small comments

File: api/src/paths/project/{projectId}/update.ts [hunk 3]
```diff
@@ -224,6 +236,37 @@ export const getPartnershipsData = async (projectId: number, connection: IDBConn
...
+
+  if (!resultFocalSpecies) {
+    throw new HTTP400('Failed to get focal species data');
+  }
+
+  if (!resultAncillarySpecies) {
+    throw new HTTP400('Failed to get ancillary species data');
+  }
+
+  return {
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/update.ts]: this should probably be:

```
return GetSpeciesData(resultFocalSpecies, resultAncillarySpecies)
```

File: api/src/queries/project/project-view-update-queries.ts [hunk 4]
```diff
@@ -33,3 +33,65 @@ export const getStakeholderPartnershipsByProjectSQL = (projectId: number): SQLSt
 
   return sqlStatement;
 };
+
+/**
+ * SQL query to get project focal species.
```
Comment (sdevalapurkar) [inline api/src/queries/project/project-view-update-queries.ts]: nit:
```suggestion
 * SQL query to get project focal species for update purposes.
```

File: api/src/queries/project/project-view-update-queries.ts [hunk 5]
```diff
@@ -33,3 +33,65 @@ export const getStakeholderPartnershipsByProjectSQL = (projectId: number): SQLSt
...
+    message: 'sql',
+    'sqlStatement.text': sqlStatement.text,
+    'sqlStatement.values': sqlStatement.values
+  });
+
+  return sqlStatement;
+};
+
+/**
+ * SQL query to get project ancillary species.
```
Comment (sdevalapurkar) [inline api/src/queries/project/project-view-update-queries.ts]: nit:
```suggestion
 * SQL query to get project ancillary species for update purposes.
```

File: app/src/components/fields/MultiAutocompleteFieldVariableSize.tsx [hunk 6]
```diff
@@ -109,7 +109,7 @@ const MultiAutocompleteFieldVariableSize: React.FC<IMultiAutocompleteField> = (p
   const { values, touched, errors, setFieldValue } = useFormikContext<IMultiAutocompleteFieldOption>();
 
   const getExistingValue = (existingValues: any[]): IMultiAutocompleteFieldOption[] => {
-    if (!values) {
+    if (!existingValues) {
```
Comment (sdevalapurkar) [inline app/src/components/fields/MultiAutocompleteFieldVariableSize.tsx:112]: question: why was this changed?

File: app/src/constants/i18n.ts [hunk 7]
```diff
@@ -48,3 +48,10 @@ export const EditLocationBoundaryI18N = {
   createErrorText:
     'An error has occurred while attempting to edit your location boundary, please try again. If the error persists, please contact your system administrator.'
 };
+
+export const EditSpeciesI18N = {
+  editTitle: 'Edit focal and ancillary species',
+  createErrorTitle: 'Error Editing Species',
+  createErrorText:
```
Comment (sdevalapurkar) [inline app/src/constants/i18n.ts]:
```suggestion
  editErrorTitle: 'Error Editing Species',
  editErrorText:
```

File: app/src/features/projects/view/components/Species.test.tsx [hunk 8]
```diff
@@ -1,35 +1,132 @@
...
+    const { asFragment } = render(
+      <Router history={history}>
+        <Species projectForViewData={{ ...getProjectForViewResponse }} codes={codes} />
+      </Router>
+    );
 
     expect(asFragment()).toMatchSnapshot();
   });
+
+  it('editing the partnerships works in the dialog', async () => {
```
Comment (sdevalapurkar) [inline app/src/features/projects/view/components/Species.test.tsx]:
```suggestion
  it('editing the species works in the dialog', async () => {
```

File: app/src/features/projects/view/components/Species.tsx [hunk 9]
```diff
@@ -37,21 +49,66 @@ const useStyles = makeStyles({
...
+          initialValues: speciesForUpdate,
+          validationSchema: ProjectSpeciesFormYupSchema
+        }}
+        onClose={() => setOpenEditDialog(false)}
+        onCancel={() => setOpenEditDialog(false)}
+        onSave={handleDialogEditSave}
+      />
+
+      <ErrorDialog
+        dialogTitle="Failed to Fetch Edit Data"
```
Comment (sdevalapurkar) [inline app/src/features/projects/view/components/Species.tsx]:
```suggestion
        dialogTitle="Failed to Fetch Species Data"
```

File: app/src/constants/i18n.ts [hunk 10]
```diff
@@ -48,3 +48,10 @@ export const EditLocationBoundaryI18N = {
   createErrorText:
     'An error has occurred while attempting to edit your location boundary, please try again. If the error persists, please contact your system administrator.'
 };
+
+export const EditSpeciesI18N = {
+  editTitle: 'Edit focal and ancillary species',
```
Comment (sdevalapurkar) [inline app/src/constants/i18n.ts]: I think we should be consistent with casing here, so either:
1) `Edit Species`
or
2) `Edit Focal and Ancillary Species`

Comment (NickPhura) [reply app/src/constants/i18n.ts]: Probably `Edit Species` as they are bundled together under a single 'Species` section

File: app/src/components/fields/MultiAutocompleteFieldVariableSize.tsx (hunk 6 above)
Comment (NickPhura) [reply app/src/components/fields/MultiAutocompleteFieldVariableSize.tsx:112]: We ran into a `cant access .includes of undefined` at one point, which put some focus on this component. 

And then, looking at this fucntion now, I don't the previous check was useful, as `values` wasn't used in this function, but exstingValues is.

File: app/src/features/projects/view/components/Species.tsx [hunk 11]
```diff
@@ -37,21 +49,66 @@ const useStyles = makeStyles({
...
+    codes
   } = props;
 
+  const history = useHistory();
+  const biohubApi = useBiohubApi();
+
   const classes = useStyles();
 
+  const [openEditDialog, setOpenEditDialog] = useState(false);
+  const [openErrorDialog, setOpenErrorDialog] = useState(false);
```
Comment (sdevalapurkar) [inline app/src/features/projects/view/components/Species.tsx]: I would change this to be consistent with how we are probably going to do it in the final version:

```
const [errorDialogProps, setErrorDialogProps] = useState<IErrorDialogProps>({
    dialogTitle: EditSpeciesI18N.editErrorTitle,
    dialogText: EditSpeciesI18N.editErrorText,
    open: false,
    onClose: () => {
      setErrorDialogProps({ ...errorDialogProps, open: false });
    },
    onOk: () => {
      setErrorDialogProps({ ...errorDialogProps, open: false });
    }
  });

  const showErrorDialog = (textDialogProps?: Partial<IErrorDialogProps>) => {
    setErrorDialogProps({ ...errorDialogProps, ...textDialogProps, open: true });
  };
```

File: app/src/features/projects/view/components/Species.tsx [hunk 12]
```diff
@@ -37,21 +49,66 @@ const useStyles = makeStyles({
...
+        onSave={handleDialogEditSave}
+      />
+
+      <ErrorDialog
+        dialogTitle="Failed to Fetch Edit Data"
+        dialogText="Could not retrieve data for editing purposes, please try again later."
+        open={openErrorDialog}
+        onClose={() => setOpenErrorDialog(false)}
+        onOk={() => setOpenErrorDialog(false)}
+      />
```
Comment (sdevalapurkar) [inline app/src/features/projects/view/components/Species.tsx]: then this will be simplified to:

```
<ErrorDialog {...errorDialogProps} />
```

**Review by sdevalapurkar — COMMENTED** [review]
need to do something like this:

```
const handleDialogEditOpen = async () => {
    let speciesResponseData;

    try {
      const response = await biohubApi.project.getProjectForUpdate(id, [UPDATE_GET_ENTITIES.species]);

      if (!response?.species) {
        showErrorDialog({ open: true });
        return;
      }

      speciesResponseData = response.species;
    } catch (error) {
      const apiError = new APIError(error);
      showErrorDialog({ dialogText: apiError.message, open: true });
      return;
    }

    setSpeciesForUpdate(speciesResponseData );

    setOpenEditDialog(true);
  };
```

File: app/src/features/projects/view/components/Species.tsx [hunk 13]
```diff
@@ -61,25 +62,52 @@ const Species: React.FC<ISpeciesProps> = (props) => {
   const classes = useStyles();
 
   const [openEditDialog, setOpenEditDialog] = useState(false);
-  const [openErrorDialog, setOpenErrorDialog] = useState(false);
+  //const [openErrorDialog, setOpenErrorDialog] = useState(false);
```
Comment (sdevalapurkar) [inline app/src/features/projects/view/components/Species.tsx]: remove this

File: app/src/features/projects/view/components/Species.tsx [hunk 14]
```diff
@@ -61,25 +62,52 @@ const Species: React.FC<ISpeciesProps> = (props) => {
...
+      setErrorDialogProps({ ...errorDialogProps, open: false });
+    },
+    onOk: () => {
+      setErrorDialogProps({ ...errorDialogProps, open: false });
+    }
+  });
+
+  const showErrorDialog = (textDialogProps?: Partial<IErrorDialogProps>) => {
+    setErrorDialogProps({ ...errorDialogProps, ...textDialogProps, open: true });
+  };
```
Comment (sdevalapurkar) [inline app/src/features/projects/view/components/Species.tsx:110]: supernit: add newline before return of the JSX

File: app/src/features/projects/view/components/Species.tsx [hunk 15]
```diff
@@ -37,21 +50,86 @@ const useStyles = makeStyles({
...
```
Comment (sdevalapurkar) [inline app/src/features/projects/view/components/Species.tsx:110]:
```suggestion
  };
  
```

File: app/src/features/projects/view/components/Species.tsx [hunk 16]
```diff
@@ -37,21 +50,86 @@ const useStyles = makeStyles({
...
+    } catch (error) {
+      const apiError = new APIError(error);
+      showErrorDialog({ dialogText: apiError.message, open: true });
+      return;
+    }
+
+    setSpeciesForUpdate(speciesResponseData);
+
+    setOpenEditDialog(true);
+  };
```
Comment (sdevalapurkar) [inline app/src/features/projects/view/components/Species.tsx:88]:
```suggestion
  };
  
```

**Review by sdevalapurkar — DISMISSED** [review]

**Review by sdevalapurkar — APPROVED** [review]
